### ISSUE TITLE
[bug] Critical Error:  Invalid memory read/write

### ISSUE THREAD

**Issue by SAbboushi** [zero-plusplus/vscode-autohotkey-debug#26]
```
#SingleInstance 

CLSID_TaskbarList := "{56FDF344-FD6D-11d0-958A-006097C9A090}"
IID_ITaskbarList  := "{56FDF342-FD6D-11d0-958A-006097C9A090}"

ComWrapperObject_TaskbarList := ComObjCreate( CLSID_TaskbarList, IID_ITaskbarList )

vPtr_ITaskbarList := ComObjValue( ComWrapperObject_TaskbarList )
ComWrapperObject_TaskbarList := ComObject( 13, vPtr_ITaskbarList )

ObjRelease( vPtr_ITaskbarList )

ExitApp()
```

To generate Critical Error:
1) Collapse the Sidebar VARIABLES Panel:
![image](https://user-images.githubusercontent.com/1964198/90959882-42b0ce00-e463-11ea-80df-1ec01ab11a3e.png)

2) Set a breakpoint on ObjRelease() statement
![image](https://user-images.githubusercontent.com/1964198/90959931-91f6fe80-e463-11ea-8412-02c9b6e0cfa6.png)

3) Start debugging.  Execution should stop at breakpoint

4) Expand the VARIABLES Panel:
![image](https://user-images.githubusercontent.com/1964198/90959972-e7331000-e463-11ea-9662-22a30f00130c.png)

Critical Error window is displayed:
![image](https://user-images.githubusercontent.com/1964198/90959985-fe71fd80-e463-11ea-8302-40795b0c7508.png)

**Comment by SAbboushi** [comment]
After further testing, this is my understanding:

```
/*  
    In my example code above, ComObjCreate() creates a COM wrapper object.  The object "wraps" the COM interface
    pointer.  The COM interface pointer is stored in the wrapper object's "ptr" property and can be retrieved using
    ComObjValue().  Deleting the COM wrapper object also deletes the COM interface which means the ptr to the COM 
    Interface is no longer valid.  ptr is now a dangling pointer i.e. pointing to an object in memory that has been 
    freed.
    
    A simpler example:
*/

Object1 := {}
Object1_address := ObjPtr( Object1 )
Object1 := ""
myObject := ObjFromPtr(Object1_address) ; this will throw error if debugger tries to inspect myObject (e.g. expand
                                        ; VARIABLES panel or hover mouse pointer over myObject)

```
YOUR SAMPLE CODE:
```
address := ObjPtr({})   ; "address" is an "unmanaged" pointer to an object.  After the object's address is assigned to 
                        ; "address", the object is deleted because there is no "known" reference to the object i.e.
                        ;   the object's reference count is zero.  
                        ; ObjPtr() does not increase the object's reference count.

;ObjRelease(address)     ; "address" is a "dangling pointer" because it is the address to an object that no longer 
                        ; exists in memory.  This can be confirmed with:
obj := ObjFromPtr( address )    ; Add a Breakpoint to ObjectRelease() statement (the last line).  
                                ; If error does not appear at breakpoint, hover over "obj" or expand VARIABLES panel.
                                ; Note: I also find that moving the mouse pointer up and down in the Editor sometimes
                                ;   triggers data inspection and thus throws the error.

; We can generate the same error by randomly choosing any valid memory address:
Not_a_valid_pointer := 12345678

ObjRelease( Not_a_valid_pointer )   ; Critical Error:  Invalid memory read/write.
                                    ; I find a different behavior though: this always throws an error even without
                                    ; trying to force debugger to inspect variable(s).  
                                    ; I wonder if that is because "address" variable is/was internally associated with
                                    ; an object or object pointer whereas "Not_a_valid_pointer" variable is not 
                                    ; /never was?

```

**Comment by zero-plusplus** [comment]
My example was funny. It is correct.
```ahk
obj := {}
address := ObjPtr(obj)
ObjRelease(address) ; error
```

In this example, the reference counter is set to 1 and it shouldn't be a dangling pointer either.
I thought ObjRelease would work correctly, but in fact it causes an error.

If you add `ObjAddRef(address)` to this example, ObjRelease does not cause the error. In this case, the reference counter becomes 2 and ObjRelease returns 1.

Hmmm. I don't know enough about it.

**Comment by SAbboushi** [comment]
I hope I understood your post so that you find my response useful.  I apologize if I misunderstood.

```
obj := {}   ; obj refcount = 1

; We can get the current reference count by doing this:
obj_refcount := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; 1

address := ObjPtr(obj)  ; ObjPtr() does not increase the object's "official" reference count.
obj_refcount1 := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; still 1

; NOTE: although address is a second reference to obj, AHK does not count this as a reference unless we use ObjAddRef()

;obj_refcount2 := ObjRelease(address)   ; NOT GOOD TO DO THIS: reduces obj refcount to 0, 
                                        ; so AHK deletes obj and throws error.

/* 
    I think this error is to be expected and that this is not good to manually reduce the reference count 
    on an AHK-managed object to zero like that: 
    AHK automatically manages AHK objects, pointers and reference count e.g.
*/
obj := {}               ; AHK automatically manages refcount; 1
obj_alias := obj        ; AHK automatically manages refcount; 2
obj_refcount3 := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; 2
obj_refcount4 := ObjAddRef(ObjPtr(obj_alias))*0 + ObjRelease(ObjPtr(obj_alias)) ; 2

obj_alias := 2  ; AHK automatically manages refcount, reference to object is released.  refcount to object reduced by 1
obj_refcount5 := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; 1

obj_address := ObjPtr(obj)  ; create an unmanaged pointer to object which means we need to manually manage it.
                            ; refcount does not change

obj_refcount6 := ObjAddRef( obj_address ) ; refcount to object increased by 1
obj_refcount7 := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; 2

obj := ""   ; AHK automatically manages refcount, reference to object is released.  refcount to object reduced by 1

; The next line (obj_refcount8) will throw an error because "obj" is no longer a reference to the object.
; But because the object's reference count is still one, the object is still stored in memory
; at "obj_address".
;obj_refcount8 := ObjAddRef(ObjPtr(obj))*0 + ObjRelease(ObjPtr(obj)) ; Error:  Expected an object but got a String.

; Note that the object is no longer "automatically" managed by AHK.
; We can pass obj_address via dllcall() or possibly use it in some other way.  But because this is an
; "unmanaged pointer to an object", we need to manually release the obj_address reference when we're finished with
; the object i.e. "ObjRelease( obj_address )"

;obj_address := ""   ; this will clear the variable, but because this is an "unmanaged pointer to an object"
                    ;   the object is not deleted from memory.  Its memory space will not be released
                    ;   and without the address, we can no longer release/recover the memory used by the object.
                    ;   This has created a "memory leak"

ObjRelease( obj_address )   ; this is correct way to release the remaining reference to the object

/*
    In summary, ObjRelease() is for use with "unmanaged" pointers to objects.  
    I don't think there is a need to use ObjRelease() unless you first use ObjAddRef().
```

**Comment by zero-plusplus** [comment]
Thanks for the explanation.
My question wasn't resolved because I didn't know that ObjRelease assumes an unmanaged object, so need to use ObjAddRef.
I thought my example was a manual replacement for the automatic management by AHK, so I thought it should work fine.

thanks!

**Comment by SAbboushi** [comment]
I am happy that my explanation was useful!
